### PR TITLE
MethodDL fixes and optimisations

### PR DESCRIPTION
**Pull request by lmoneta**
Apply several improvements and fixes in MethodDL , mainly: 

- Use now Float instead of Double. This makes the network 2 times faster without loosing performances in quality of results
- Compute now the regularisation term in the loss once/epoch when evaluating 
- Optimise the MultiThread execution on CPU by setting a minimum number of elements to execute per task 
- Optimise evaluation of the networks, by implementing GetMvaValues in MethodDL. This reduces significantly the time spent for network testing and evaluating in a TMVA session

### CODE REVIEW

**Comment by lmoneta** [comment]
@phsft-bot build !

File: tmva/tmva/CMakeLists.txt [hunk 1]
```diff
@@ -76,7 +76,7 @@ endif()
 
 #---Handle BLAS dependent code. -----------------
 if(BLAS_FOUND AND imt)
-   message(STATUS "Using TMVA-DNN with BLAS installation") 
+   message(STATUS "Using TMVA-DNN with BLAS installation: ${BLAS_LIBRARIES}")
```
Comment (ashlaban) [inline tmva/tmva/CMakeLists.txt]: Does this print the path to the installed library (which I think would be very helpful when debugging!)

File: tmva/tmva/inc/TMVA/DNN/Net.h [hunk 2]
```diff
@@ -189,8 +189,8 @@ TNet<Architecture_t, Layer_t>::TNet(size_t batchSize,
       AddLayer(other.GetLayer(i).GetWidth(),
                other.GetLayer(i).GetActivationFunction(),
                other.GetLayer(i).GetDropoutProbability());
-      fLayers[i].GetWeights() = (TMatrixT<Double_t>) other.GetLayer(i).GetWeights();
-      fLayers[i].GetBiases()  = (TMatrixT<Double_t>) other.GetLayer(i).GetBiases();
+      fLayers[i].GetWeights() = (TMatrixT<Scalar_t>) other.GetLayer(i).GetWeights();
```
Comment (ashlaban) [inline tmva/tmva/inc/TMVA/DNN/Net.h]: The type parameter is in some places called `Scalar_t`, in some places `AFloat` and in some places `AReal`. Since we're changing this anyway, maybe we want to ensure some consistency.

File: tmva/tmva/inc/TMVA/DNN/Architectures/Cuda/CudaMatrix.h [hunk 3]
```diff
@@ -161,7 +161,7 @@ class TCudaMatrix
    TCudaDeviceReference<AFloat> operator()(size_t i, size_t j) const;
 
    void Print() const { 
-      TMatrixT<Double_t> mat(*this); 
+      TMatrixT<AFloat> mat(*this);
```
Comment (ashlaban) [inline tmva/tmva/inc/TMVA/DNN/Architectures/Cuda/CudaMatrix.h]: See comment on `AFloat` vs `Scalar_t` vs `AReal`.

File: tmva/tmva/src/DNN/Architectures/Cpu/Regularization.cxx [hunk 4]
```diff
@@ -21,6 +21,147 @@ namespace TMVA
...
+#else
+      for (size_t i = 0;  i < nElements; i+=nSteps)
+         f(i);
+#endif
+   } else {
+      f(0);
+   }
+}
+
+#if 0  // old implementation (to be removed)
```
Comment (ashlaban) [inline tmva/tmva/src/DNN/Architectures/Cpu/Regularization.cxx]: Why not remove the implementation now? It will remain in the git history if we want to bring it back.

File: tmva/tmva/src/MethodDL.cxx [hunk 5]
```diff
@@ -340,6 +344,8 @@ void MethodDL::ProcessOptions()
          settings.regularization = DNN::ERegularization::kL1;
       } else if (regularization == "L2") {
          settings.regularization = DNN::ERegularization::kL2;
+      } else {
```
Comment (ashlaban) [inline tmva/tmva/src/MethodDL.cxx]: This is fine, but a change to consider is making a separate option for each regulariser (in principle one can use them at the same time).
E.g. `L2_Regularisation=1.5:L1_Regularisation=0.`. The default of =0. would disable the regulariser.
Right now it is unclear to me how you control the strength of the regularisation. (Is this `WeightDecay`?)

File: tmva/tmva/src/MethodDNN.cxx [hunk 6]
```diff
@@ -1095,7 +1095,7 @@ void TMVA::MethodDNN::TrainCpu()
          p = 1.0 - p;
       }
       net.SetDropoutProbabilities(dropoutVector);
-      //net.SetDropoutProbabilities(settings.dropoutProbabilities);
+      //net.SetDropoutProbabilities(settings.dropout<Probabilities);
```
Comment (ashlaban) [inline tmva/tmva/src/MethodDNN.cxx]: Spurious <

File: tmva/tmva/src/MethodDL.cxx [hunk 7]
```diff
@@ -1107,13 +1142,17 @@ void MethodDL::TrainDeepNet()
...
+            << " regularization " << (char) settings.regularization 
+            << " minimum error = " << minTestError
+            << Endl;
       if (!fInteractive) {
          std::string separator(62, '-');
          Log() << separator << Endl;
          Log() << std::setw(10) << "Epoch"
                << " | " << std::setw(12) << "Train Err." << std::setw(12) << "Test Err." 
-               << std::setw(12) << "t(s)/epoch" << std::setw(12)  << "Eval t(s)"
+               << std::setw(12) << "t(s)/epoch" << std::setw(12)  << "t(s)/Loss"
```
Comment (ashlaban) [inline tmva/tmva/src/MethodDL.cxx]: Is this correct? (Time per loss, what does it mean? Sorry if I'm tired and just don't see it!)

File: tmva/tmva/CMakeLists.txt (hunk 1 above)
Comment (lmoneta) [reply tmva/tmva/CMakeLists.txt]: Yes, it prints that.

And thanks for having seen the mixmcx file committed by mistake

File: tmva/tmva/src/DNN/Architectures/Cpu/Regularization.cxx (hunk 4 above)
Comment (lmoneta) [reply tmva/tmva/src/DNN/Architectures/Cpu/Regularization.cxx]: yes, correct , I'll do it

File: tmva/tmva/src/MethodDL.cxx (hunk 5 above)
Comment (lmoneta) [reply tmva/tmva/src/MethodDL.cxx]: yes weight decay controls the strength !
I am not sure I understand your comment about  L2_Regularisation=1.5:L1_Regularisation=0..
Now you have a string defining the regularisation, can be "None","L1","L2"

File: tmva/tmva/src/MethodDL.cxx (hunk 7 above)
Comment (lmoneta) [reply tmva/tmva/src/MethodDL.cxx]: It means the time to compute the Loss function for both training and testing data.
For a fixed interval of epochs (defined in testRepetitions) we compute the test error on both 
test and training data set. Maybe we should call t(s)/Test ?

File: tmva/tmva/src/MethodDL.cxx (hunk 5 above)
Comment (ashlaban) [reply tmva/tmva/src/MethodDL.cxx]: I'm just saying that in e.g. xgboost the configuration options are
`xgb.train(..., l2_reg=1., l1_reg=0.2, ...)`, that is they are separate, and maybe we would want to have a similar interface.

**Comment by ashlaban** [comment]
@phsft-bot build!

**Comment by lmoneta** [comment]
@phsft-bot build !

**Comment by ashlaban** [comment]
Waiting for #2438 through #2441 before merging.
